lib/elixir/lib/kernel/special_forms.ex

```diff
@@ -1291,6 +1291,39 @@ defmodule Kernel.SpecialForms do
   defmacro for(args)
 
   @doc """
+  Used to combine matching clauses.
+
+  Let's start with an example:
+
+      iex> opts = %{width: 10, height: 15}
+      iex> with {:ok, width} <- Map.fetch(opts, :width),
+      ...>      {:ok, height} <- Map.fetch(opts, :height),
+      ...>   do: {:ok, width * height}
+      {:ok, 150}
+
+  If all clauses match, the `do` block is executed, returning its result.
+  Otherwise the chain is aborted and a non-matched value is returned:
+
+      iex> opts = %{width: 10}
+      iex> with {:ok, width} <- Map.fetch(opts, :width),
+      ...>      {:ok, height} <- Map.fetch(opts, :height),
+      ...>   do: {:ok, width * height}
+      :error
+
+  Similarly to `for`, variables bound inside `with` won't leak.
+  Also similar to `for`, `with` allows "bare expressions":
+
+      iex> opts = %{width: 10, height: 15}
+      iex> with {:ok, width} <- Map.fetch(opts, :width),
+      ...>      double_width = width * 2,
+      ...>      {:ok, height} <- Map.fetch(opts, :height),
+      ...>   do: {:ok, double_width * height}
+      {:ok, 300}
+
```

Comment:
I will be good to add an example showing that variables won't leak, setting either `width` or `double_width` before `with/1` it, and printing a result at the end